lib/python/treadmill_aws/dns_sync.py

```diff
@@ -0,0 +1,228 @@
+"""Syncronize DNS data with Zookeeper file system mirror.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import glob
+import io
+import logging
+import os
+
+import six
+
+from treadmill import utils
+from treadmill import yamlwrapper as yaml
+
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_WEIGHT = 10
+DEFAULT_PRIORITY = 10
+
+
+class DnsSync(object):
+    """Syncronizes DNS with Zk mirror on disk."""
+
+    def __init__(self, ipaclient, cell, zone, fs_root, scopes):
+        self.ipaclient = ipaclient
+        self.cell = cell
+        self.fs_root = os.path.realpath(fs_root)
+
+        self.state = None
+        self.servers = set()
+        self.scopes = {
+            'cell': cell,
+        }
+        self.scopes.update(scopes)
+
+        self.zone = zone
+        self.zones = ['.'.join([scope, scope_name, self.zone])
+                      for scope_name, scope in six.iteritems(self.scopes)]
+
+    def _srv_rsrc(self, name, scope, proto, endpoint, hostport):
+        """Return tuple of resource endpoint/payload."""
+        host, port = hostport.split(':')
+        if scope not in self.scopes:
+            _LOGGER.warning('Unsupported scope: %s', scope)
+            return None
+
+        if proto not in ['tcp', 'udp']:
+            _LOGGER.warning('Unsupported proto: %s', proto)
+            return None
+
+        zone = '.'.join([self.scopes[scope], scope, self.zone])
+        return (
+            '_{endpoint}._{proto}.{name}.{zone}'.format(
+                endpoint=endpoint,
+                proto=proto,
+                name=name,
+                zone=zone),
+            '{weight} {priority} {port} {host}.'.format(
+                weight=DEFAULT_WEIGHT,
+                priority=DEFAULT_PRIORITY,
+                port=port,
+                host=host
+            )
+        )
+
+    def _current_records(self, zone):
+        """Return all records found in DNS for the given cell."""
+        _LOGGER.info('Query state for: %s', zone)
+
+        result = self.ipaclient.get_dns_record(self.zone)
+        if result.get('error'):
+            raise Exception(result.get('error'))
+
+        all_records = result['result']['result']
+        current = set()
+        for rec in all_records:
+            if 'srvrecord' not in rec:
+                continue
+
+            srv_rec_name = rec['idnsname'][0]
+            for srv_rec in rec['srvrecord']:
+                _wei, _prio, port, host = srv_rec.split()
+
+                # skip records that do not belong to the cell.
+                host = host.rstrip('.')
+                if host not in self.servers:
```

Comment:
i cannot forgo the check as these are by design multi-cell records. and we will have multiple cells in aws.

hygiene is a concern, and it is concern internally too. that said, the flow is probably - deco server first, it shutsdown, srv records deleted (because it is still part of the cell then down) and then there is gc process that removes deleted instances from the cell.

in any case, this gc process is not finalized yet, but we can't forgo the check here.